### DISCUSSION TITLE
Real-time energy monitoring tweak

### DISCUSSION THREAD

**Discussion by hepcat72** [arendst/Tasmota#10391] [Q&A] [answered]
For some reason, I can't copy message links from Discord, so to summarize what leads me to opening this discussion...

I am trying to use wattage changes as automation triggers.  My goal is to flip a dumb switch on a device plugged into a sonoff and make other things happen.  I have tried both PowerDelta and a combination of PowerLow/PowerHigh rules, but there's a delay of around 2 seconds regardless of the method for detecting the wattage change.  I have a Sonoff S31 15A.

[For further background, I have also tried a Currant brand outlet and I could set the same threshold on it and it detected the wattage change with almost no delay.  The only problem with that was they didn't have a means of getting the triggers into Node RED.]

So I've been studying the code to see if I can tweak anything to make detection of wattage changes closer to real-time (both via PowerDelta and low/high rules thresholds - which, BTW appear to be only negligibly different), ideally a fraction of a second. I found where Energy.active_power is calculated and set and the code looks like it *should* report power changes within 1 second. My initial hunch was that Energy.power_history might have been mis-handled (since my observation:

https://drive.google.com/open?id=1_C4OFqSyKh6MxeL2yJqyaPd11vWmHUYI

was that changes were detected in about 2 seconds and power_history contains readings from 2 seconds before "now" [note, I move the mouse when I flip the switch in the above video: up=on, down=off])... but it all looks right.

My other hunch is that the power readings are averaged over a window, sampled once a second, thus causing an immediate change to be caught on a delay (>1s). My oberservations in the web GUI support this.  When I flip the switch, it goes from 1W to 6W (1s later), then to the target 11W (1s after that).

I have a question.  Power readings are obtained from the sensor and active_power is calculated in a loop here:

https://github.com/arendst/Tasmota/blob/dcbb3f1e0ea2141641c5d4ed5ed8f6d793d26a68/tasmota/xnrg_02_cse7766.ino#L145

It must be those values that possibly have the windowed-average I'm hoping to find.  However, I can't determine how frequently that while loop cycles.  Is it potentially looping faster than once a second and is that looping independently of the loop that checks the margins:

https://github.com/arendst/Tasmota/blob/628f17de8c91bade92edefe3b1b4d544e644710d/tasmota/tasmota.ino#L379

I guess even if that loop runs once a second, it's the sender that would be implementing the windowed average.  Otherwise, the while loop would catch the change immediately...

I guess I'm just thinking out loud here.

Oh, BTW, the link to the PDF in the comments is stale:

https://github.com/arendst/Tasmota/blob/dcbb3f1e0ea2141641c5d4ed5ed8f6d793d26a68/tasmota/xnrg_02_cse7766.ino#L28

I found the doc here:

https://dl.itead.cc/S31/CSE7766.pdf

**Comment by Frogmore42** [comment]
You might want to back up and explain more about why you are choosing this method.  What are you really trying to do?
There are reasons that using power utilization might be the best solution, but normally it is better to use an actual switch to a device.  Some of the Shelly devices can take a high voltage switch input and can run Tasmota.  You will get very quick response from that.

When you put power measuring circuitry in between, you get the delay of measuring power accurately (averaging is usually involved), and the delay of reporting, which is not instant.

Looking at the code:

1. CseSerialInput gets called whenever there is serial data available from the chip.  The datasheet is not really clear on how often this will be, but it does have this:
```
Note: Maximum measurement is about 16s for current cycle, voltage cycle and active
power cycle. It will prompt that the measuring range of corresponding cycle is exceeded
after exceeded, as shown in Table 3-3.
```
Later it has this:
```
The unit for voltage cycle, current cycle and power cycle is us. When signal cycle
is larger than 1s, the signal cycle will be outputted to the user immediately, and it will
prompt through Adj.4(5 or 6) that it is an incomplete cycle. (For example, if signal
cycle is 3.5s, 1s signal cycle will be outputted to the user when the accumulated cycle
counts to 1s, and 1.5s signal cycle will be outputted to the user when the accumulated
cycle counts to 1.5s instead the message is received by the user after reaching 3.5s so
that the response speed is improved. When the signal amplitude to be measured
declines, the signal cycle will increase.)
```
That says to me that, at best you are going to get a few samples per second for normal values of power.
2. Once CseSerialInput gets a complete transmission from the chip, it calls CseReceived()
3. CseReceived() interprets the data and sets some variables with the high-level information
4. CseEverySecond() gets called (every second) to validate the data and then calls EnergyUpdateToday()
5. EnergyUpdateToday() takes the data and updates its variables
6. EnergyEverySecond() takes the variables and does some checks and then calls EnergyMarginCheck()
7. EnergyMarginCheck() does the actual check you want

I am not really sure the order of the two modules (Energy (xdrv03) and CSE7766 (xnrg_02)).  But, there is a lot of code there.  It is NOT designed to be used for things that need to have a near instant response (as you have discovered, it can take some seconds for the action to occur).

The WebUI updates 1/sec.  When there is 0/low power utilization the CSE7766 takes more than a second to calculate it.  So, when the power changes it will take it some time to figure this out.  Power measurement takes time, you probably won't be able to make it work much faster.  If you need real time support, use an actual switch input.

**Reply by hepcat72** [reply]
I just lost my reply. Dumb phone. Anyway...

> When you put power measuring circuitry in between, you get the delay of measuring power accurately (averaging is usually involved), and the delay of reporting, which is not instant.

To summarize my lost response. I just want to get the delay to under a second. I know it can be done, though I'm not sure it can be done with a Sonoff, but in principle, it is possible.  I just don't know how much work is involved. I know it's possible, because the Currant outlet does it. I bought one and played around with it. You can set a wattage threshold and get a notification when the wattage goes above or below that threshold, and it was really fast, nearly immediate. My only problem with the Currant was that it only works in the SmartThings ecosystem and none of my other devices or platforms was compatible with that, so I returned it and bought a Sonoff.

As to why I'm doing this... I'm doing it for fun. I like to tinker. Thanks for the details. I'll look things over and see what I can figure out.

I'd mentioned this in Discord, but I hope to implement a setting that would allow one to tweak the responsiveness of the energy monitoring triggers (I.e. if there's some sort of trade-off, it can be a configurable mode). So no worries.

I'll send a link to the Currant in a bit here if you want to try it out.

**Reply by hepcat72** [reply]
Here's the Currant:

https://www.amazon.com/dp/B07HKVRGNN/ref=cm_sw_r_cp_api_glc_i_jIO8FbJ23HHYN

Note that their default threshold in their app is 0W, but if you contact their support, they can set a custom threshold for you. They told me their next app version should allow the user to set a custom threshold.

You can add it to the SmartThings app though and in there you can set a custom threshold. And like I said, their detection of the wattage change happened in a small fraction of a second, near immediate.

Interestingly (tangential side story here) the Currant outlet I bought initially had a problem. It was sending the notifications minutes after the wattage change. Turned out it was a bug in their firmware. Their software is supposed to have a component to map around bad storage spots, but my specific outlet had a bad spot in a header and the component that read the header had a bug in the re-mapping routine. They were able to debug, fix, and push out a new firmware to my outlet that fixed it.

**Reply by hepcat72** [reply]
I'll need to reread your comments again when my wife's not bugging me to come to bed, but I'm glad to hear that the averaging I expect, is somewhere in the code. Incidentally, my intent is to be able to map the noise. If a stable average can be established, it should be possible to see a new value that rises above the noise and act instantly on it. I'm not a statistician, but I work with noisy data a lot. Should be totally doable, though I've never looked at wattage data before, so we'll see.

**Reply by hepcat72** [reply]
OK.  I have digested your response further and have looked more at the documentation and the code.  I'm not yet quite sure about this, but I think I may have found an issue that if addressed, could very well make energy changes more immediate.  I need to look into it more, but have spent enough time on it today.

The factors which make me as yet unsure are:

1. I haven't traced the code through yet to determine when precisely this code is called/executed:

https://github.com/arendst/Tasmota/blob/dcbb3f1e0ea2141641c5d4ed5ed8f6d793d26a68/tasmota/xnrg_02_cse7766.ino#L90-L96

2. The wording of the documentation (specifically, the second section you referenced) is somewhat ambiguous.

However, regarding the documentation, I infer from other sections, the following:

My interpretation of:

> The unit for voltage cycle, current cycle and power cycle is us. When signal cycle is larger than 1s, the signal cycle will be outputted to the user immediately, and it will prompt through Adj.4(5 or 6) that it is an incomplete cycle. (For example, if signal cycle is 3.5s, 1s signal cycle will be outputted to the user when the accumulated cycle counts to 1s, and 1.5s signal cycle will be outputted to the user when the accumulated cycle counts to 1.5s instead the message is received by the user after reaching 3.5s so that the response speed is improved. When the signal amplitude to be measured declines, the signal cycle will increase.)

is that energy readings are transmitted on a regular schedule (i.e. the "signal cycle"), but that "signal amplitude" can change that schedule.  And I am positing that perhaps, when that amplitude is above some threshold, the information is transmitted "immediately", before the completion of a normal "signal cycle" (i.e. an "incomplete" "signal cycle").  And I believe that that reporting frequency ("signal cycle") is made less frequent (i.e. "an increased cycle") when the signal amplitude (e.g. "power") is lower (e.g. "When the signal amplitude to be measured declines, the signal cycle will increase").  So, with this, I think there might be 2 meanings of "cycle" that are confusing to tease apart because of the poor grammar & punctuation in the document:

- signal cycle (i.e. reporting time/schedule)
- signal amplitude cycle (e.g. power cycle, i.e. a power measurement)

The text is very confusing, but reading between the lines, seems to indicate that a cycle (which I infer to mean reporting frequency ("signal cycle")) can change (e.g. "the signal cycle will increase").  And I think that it is saying that normally, energy measurements are delivered (sic "outputted") to the receiver every 1s (which I infer to mean 1 second) (if that is set a frequency that is configured somewhere).  But, it can deliver the data "immediately", or less frequently as usage changes.

And I posit generally, that the intent of the above is to convey that when energy usage is either stable (i.e. not changing much), or perhaps "low", the period ("signal cycle") is increased.  When it changes significantly, there is an immediate signal, which is indicated by an "incomplete cycle" indicated in Adj - and should be accompanied by a different power coefficient necessary to accurately compute the energy reading.  Here's support for that hypothesis:

> Adj.4: When power signal cycle is larger than 1s, present cycle counting time will be transmitted to the user. This flag bit is used to identify whether the power cycle outputted at this moment is complete power cycle or power cycle counting larger than 1s.

> Adj.4=1 indicates power cycle of serial port outputted is a complete cycle;
> Adj.4=0 indicates power cycle of serial port outputted is an incomplete cycle;

I suppose it's possible that the intent here could go the other way.  This all could be about an energy report that has been delayed, and that it will be delivered immediately if the elapsed time is longer than the configured schedule (e.g. every 1s)... Or... it could be a combination - and serve both purposes.

However, in either case, I think that the purpose of Adj is to know when the calculation of an energy level should use a different calibration, based on a different power coefficient than the schedule dictates.  It can be a larger coefficient (when the reporting time is longer than 1s) or it can be a lower coefficient (when data is transmitted "immediately" because signal amplitude is high).

The code suggests that only the former case is supported, because it interprets Adj.4=0 as an "invalid" power reading:

https://github.com/arendst/Tasmota/blob/dcbb3f1e0ea2141641c5d4ed5ed8f6d793d26a68/tasmota/xnrg_02_cse7766.ino#L108

Instead, it could simply compute the power using the accompanying power coefficient.

Of course, I could be wrong about all this, but given this hypothesis, I think that it's possible that the reported power values are possibly using a stale calibration/power coefficient when measured power changes, which is why the values gradually increase over the course of 2-3 seconds.

....MAYBE - you guys know way more about this than I do, so I don't expect you to do anything about this.  I'm just trying to develop a plan to potentially improve the response time of detecting wattage changes, and I'm just providing my hypothesis to you to see what you have to say, which could inform how I proceed with my experiment.

**Reply by Frogmore42** [reply]
While I don't want to discourage anyone from trying to make things better, I am not sure how successful this effort will be.  The term "polishing a turd" came to mind and I discovered this video from Myth Busters that explores that idea: https://youtu.be/yiJ9fy1qSFI

If you are going down this path, you will need to be very careful.  The power measuring IC on the S31 (and all of these devices) is connected directly to the AC line.  You must use galvanic isolation to prevent from damaging your test equipment and/.or self.  If you don't already know how to do this, I would not suggest learning for this project.

To solve your original problem, using a switch to control something else, I would suggest getting one or more of these: https://shelly.cloud/knowledge-base/devices/shelly-i3/#pinout

They are cheap and can be hooked up directly to the line safely.  They should be able to run Tasmota and make it easy and fast to achieve what you want.

**Reply by hepcat72** [reply]
I fail to understand what your resistance is to this endeavor, but thanks for your words of caution.  Like I said, there is at least 1 smart outlet out there that *can* detect wattage changes in a fraction of a second (near instant). And given the poor wording in the documentation, and language that suggests mechanisms for improved response time, and the fact that tasmota was not designed with fast energy change response times, it seems entirely possible that a minor change in the timing of the computation of the calibration value could lead to higher resolution of wattage values.

There's a significant difference between the solution you propose and the one I'm advocating.  Your solution involves non-trivial effort if you want to change the controlling device.  With my solution, you can swap out any device with a power switch and use it as a (semi-)smart switch (i.e. "reporting" switch).

Unless you can prove that the S31 hardware is sending window-averaged values (i.e. the calculation is hard-wired into the board), I believe that it's entirely possible to implement this change, though it's also entirely possible that I'm not up to the task.

Regarding the risks, any mistake is possible, but the only change I'm thinking about is including a re-calculation of the calibration value when ADJ.4 shows that an adjustment is indicated.  I will follow the code to see what side-effects that could have, but currently, it seems like it's just for reporting, and does not affect what happens electrically.  Again, I could be wrong, so I appreciate the words of caution.

**Reply by hepcat72** [reply]
Actually, I don't even need to alter the saved calibration value. I can just put in an else that uses a temporary value.

**Reply by Frogmore42** [reply]
I am happy that you are looking at this.  I hope that you are successful.  If you are happy spending your time investigating this, that is great.  That is what makes open source so great.  You are interested in something that is important to you.  You share that passion with the world and probably someone will benefit from your work.

Most people just want a solution to their problem.  Energy monitoring circuits are not designed for controlling things, but that doesn't mean they can't be used for it.  I am looking at the Shelly 2.5.  It has a real energy monitoring IC with a real datasheet and supports 2 channels for energy values.  Looking at the datasheet, it should be possible to query the IC frequently (maybe even every 50ms).  The IC in the S31 does not have this capability.  That is what my comment about polishing a turd is about.  The S31 is fine for energy monitoring, but, as you have discovered, it is not great for getting information quickly.  The IC in the POW 1 is much worse with measuring low power levels and takes several seconds to get an answer at low wattage utilization.

**Reply by Frogmore42** [reply]
My comment about the risks is during debugging, it is generally necessary to have a serial connection to the device to see what it is doing.  It also typically requires downloading new code to the device multiple times.  Doing this safely is important.

**Reply by hepcat72** [reply]
I saw that there is a commented out debug statement in the energy monitoring code.  I wondered if I would see that in the console if I uncommented it...  Maybe it goes to the serial connection you're referring to(?).

But theoretically, I should be able to see if any change works just by watching the energy values in the web interface, since (when I flip the switch on the device plugged into the S31) it goes from 1 to 6 to 11 watts in 0 to 1 to 2 seconds of elapsed time respectively.

And again, it's entirely possible (perhaps even likely) that I'm not interpreting the verbiage in the doc correctly and that no change I make will have the desired affect.  It all hinges on whether it reports a signal early, as opposed to just late (which is how I interpret the code was written to consume).

Anyway, I have so many personal projects going, I expect my progress to be *really* slow - like measured in months.  So far, I've just been looking into it on weekend mornings while having coffee.

**Comment by arendst** [comment, marked as answer]
If you want to see quick response from a power switch flick you should start reading the response from the power management chip.

In the case of your S31 enable ``weblog 4`` and study the data received from the chip. In there you would like to see rapid changes if you flick the switch. If no rapid changes are reported all other software changes are useless.

**Reply by hepcat72** [reply]
Wow, thanks.  That's great!  I'll definitely put this on the top of my stack.  Might save me from a bunch of wasted effort.

**Reply by hepcat72** [reply]
I played around with weblog 4 today.  Definitely helpful!

I spent the day plotting at this data (normalized) while flipping the switch.  The coefficient doesn't change as I had inferred, however, looking at the other data and reading page 20 of the docs a few times, I believe I see a solution to the smoothed wattage data.  I believe the solution lies in the V1R and V2R values when adj indicates the energy data is "invalid".  It looks like 2 or 3 bits are flipping in ck that can be used to determine V1R and V2R.  I did some manual adjustments based on the formulas and it looks like the wattage can be closer to "real" when  V1R and V2R are either 2 or 1, as suggested in the doc on page 20.  So far, it's just a hunch, but there are definite jumps in the value of ck that could suggest a bit flip at just the times when the V1R and V2R adjustments are needed.  The doc suggests that V1R and V2R are not needed when the calibration is good, and if I infer that ADJ indicating an "invalid" energy reading to mean that the calibration is not good, and I figure out which bits in ck are the important ones, I think that using V1R/V2R as each either 1 or 2 in the calculation, the wattage values will instantly be more accurate.

However, regardless of what I do, there's nothing in the data that suggests I can get the switch flip event to consistently under a second (unless I change the reporting frequency).  All the data stays level for a little over a second, regardless of what I do...  I've been using a low powerhigh, and high powerlow, and a low powerdelta, which seems to make a quantum difference in the response time.  However, if I can figure out V1R and V2R, that issue would go away.

Here's the data that I think is relevant.  Everything is normalized via min/max from 0 to 100:

<img width="1099" alt="power_adj_ck" src="https://user-images.githubusercontent.com/2300532/104832359-8c5ce380-585e-11eb-99b8-6386cd58eb98.png">

The purple line shows seconds step-wise.  I added 3 type of spikes for events in the log.  The yellow is a nonsense command I entered in the log at the exact time I flipped the switch.  The navy and red spikes are the power threshold rule and the powerdelta events respectively.  Green is the power cycle value from the mailbox.  Black is my calculated active power.  Light blue is the ADJ value (up=valid, down=invalid).  And the brown line is the ck value, which I think may be used to extract values for V1R and V2R via bit flip.

Anyway, that's enough for this weekend.  I'll play some more with it next weekend.
